sdks/java/io/azure/src/test/java/org/apache/beam/sdk/io/azure/blobstore/AzfsResourceIdTest.java

```diff
@@ -0,0 +1,308 @@
+        AzfsResourceId.fromUri("azfs://account/container/tmp/bb/")
+            .resolve("azfs://account/container/tmp/aa", RESOLVE_FILE));
+
+    // Tests container with no ending '/'.
+    assertEquals(
+        AzfsResourceId.fromUri("azfs://account/my-container/tmp"),
+        AzfsResourceId.fromUri("azfs://account/my-container").resolve("tmp", RESOLVE_FILE));
+
+    // Tests path with unicode
+    assertEquals(
+        AzfsResourceId.fromUri("azfs://account/container/输出 目录/输出 文件01.txt"),
+        AzfsResourceId.fromUri("azfs://account/container/输出 目录/")
+            .resolve("输出 文件01.txt", RESOLVE_FILE));
+  }
+
+  @Test
+  public void testResolveInvalidInputs() {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Cannot resolve a file with a directory path: [tmp/]");
+    AzfsResourceId.fromUri("azfs://account/my_container/").resolve("tmp/", RESOLVE_FILE);
+  }
+
+  @Test
+  public void testResolveInvalidNotDirectory() {
+    ResourceId tmpDir =
+        AzfsResourceId.fromUri("azfs://account/my_container/").resolve("tmp dir", RESOLVE_FILE);
+
+    thrown.expect(IllegalStateException.class);
+    thrown.expectMessage(
+        "Expected this resource to be a directory, but was [azfs://account/my_container/tmp dir]");
+    tmpDir.resolve("aa", RESOLVE_FILE);
+  }
+
+  @Test
+  public void testS3ResolveWithFileBase() {
+    ResourceId resourceId = AzfsResourceId.fromUri("azfs://account/container/path/to/file");
+    thrown.expect(IllegalStateException.class);
+    resourceId.resolve("child-path", RESOLVE_DIRECTORY); // resource is not a directory
+  }
+
+  @Test
+  public void testResolveParentToFile() {
+    ResourceId resourceId = AzfsResourceId.fromUri("azfs://account/container/path/to/dir/");
+    thrown.expect(IllegalArgumentException.class);
+    resourceId.resolve("..", RESOLVE_FILE); // '..' only resolves as dir, not as file
+  }
+
+  @Test
+  public void testEquals() {
+    AzfsResourceId a = AzfsResourceId.fromComponents("account", "container", "a/b/c");
+    AzfsResourceId b = AzfsResourceId.fromComponents("account", "container", "a/b/c");
+    assertEquals(a, a);
+    assertEquals(a, b);
+
+    b = AzfsResourceId.fromComponents(a.getAccount(), a.getContainer(), "a/b/c/");
+    assertNotEquals(a, b);
+    assertNotEquals(b, a);
+
+    b = AzfsResourceId.fromComponents(a.getAccount(), a.getContainer(), "x/y/z");
+    assertNotEquals(a, b);
+    assertNotEquals(b, a);
+
+    b = AzfsResourceId.fromComponents(a.getAccount(), "other-container", a.getBlob());
+    assertNotEquals(a, b);
+    assertNotEquals(b, a);
+
+    b = AzfsResourceId.fromComponents("other-account", a.getContainer(), a.getBlob());
+    assertNotEquals(a, b);
+    assertNotEquals(b, a);
+
+    assertEquals(
+        AzfsResourceId.fromUri("azfs://account/container"),
+        AzfsResourceId.fromUri("azfs://account/container/"));
+  }
+
+  @Test
+  public void testFromComponents() {
```

Comment:
Please also test when `blob` is null and when `blob` is empty.

Reply:
Done, thanks for bringing this to my attention.